### PR TITLE
Automatically remove delegations on user deactivation

### LINKED ISSUE #1305: Automatically remove delegations on user deactivation
This requires #1304 being implemented first.

Being a delegate or CC user should not prevent a UserProfile from being deleted.
When a user is deleted (or set as inactive, manually or via bulk deletion), the user should be removed from all other users' delegates and CC users. A message should inform about all removals that happened.

### PR DESCRIPTION
**Pull request by JannisBerndt**
Fixes #1305

### CODE REVIEW

File: evap/locale/de/LC_MESSAGES/django.po [hunk 1]
```diff
@@ -34,7 +34,7 @@ msgstr "Name (Deutsch)"
 msgid "Name (English)"
 msgstr "Name (Englisch)"
 
-#: evap/contributor/forms.py:54 evap/staff/forms.py:318
+#: evap/contributor/forms.py:54 evap/staff/forms.py:319
```
Comment (JannisBerndt) [inline evap/locale/de/LC_MESSAGES/django.po:37]: Regarding the translation: I ran `./manage.py makemessages -l de` which updated very many code line refernces that got changed sometime in the past. Do you prefer manually adding the new translations or is the big "update" done somewhere else once in a while?

Comment (janno42) [reply evap/locale/de/LC_MESSAGES/django.po:37]: we do the translations on every update of the release branch. so you only have to make sure that all strings are marked for translation in views (`_("translate this")`) and templates (`{% trans 'translate this' %}`) and i'll do the rest. please do not commit any changes on the `django.po` file.

File: evap/staff/templates/staff_user_index.html [hunk 2]
```diff
@@ -64,8 +64,7 @@
                                 ></button>
                             {% else %}
                                 <button type="button" disabled class="btn btn-sm btn-danger" data-toggle="tooltip" data-placement="left"
-                                    title="{% blocktrans %}This user contributes to an evaluation (or is a delegate or CC user of
-                                    such a person), participates in an evaluation whose participations haven't been archived yet or has special
+                                    title="{% blocktrans %}This user contributes to an evaluation, participates in an evaluation whose participations haven't been archived yet or has special
```
Comment (janno42) [inline evap/staff/templates/staff_user_index.html:67]: please have a look at the line breaks, they are a bit mixed up here. if easily possible, lines should not be longer than 120 characters.

File: evap/staff/templates/staff_user_index.html [hunk 3]
```diff
@@ -82,7 +81,7 @@
 {% block modals %}
     {{ block.super }}
     {% trans 'Delete user' as title %}
-    {% trans 'Do you really want to delete the user <strong id="data-label"></strong>?' as question %}
+    {% trans 'Do you really want to delete the user <strong id="data-label"></strong>?<br><strong id="data-label"></strong> will also be removed from every user having <strong id="data-label"></strong> as a delegated or CC-user.' as question %}
```
Comment (janno42) [inline evap/staff/templates/staff_user_index.html:84]: please change to `{% trans 'Do you really want to delete the user <strong id="data-label"></strong>?<br />This person will also be removed from every other user having this person as a delegated or CC-user.' as question %}`

File: evap/staff/tools.py [hunk 4]
```diff
@@ -292,3 +301,21 @@ def find_next_unreviewed_evaluation(semester, excluded):
         .filter(contributions__textanswer_set__state=TextAnswer.State.NOT_REVIEWED) \
         .annotate(num_unreviewed_textanswers=Count("contributions__textanswer_set")) \
         .order_by('vote_end_date', '-num_unreviewed_textanswers').first()
+
+
+def remove_user_from_represented_and_ccing_users(user, ignored_users=None, test_run=False):
+    remove_messages = []
+    ignored_users = list() if ignored_users is None else ignored_users
```
Comment (janno42) [inline evap/staff/tools.py:308]: you could write `ignored_users = ignored_users or []` instead

File: evap/staff/tools.py [hunk 5]
```diff
@@ -182,6 +188,9 @@ def bulk_update_users(request, user_file_content, test_run):
         messages.info(request, _('No data was changed in this test run.'))
     else:
         with transaction.atomic():
+            for user in deletable_users + users_to_mark_inactive:
+                # this has to be done in an additional for-loop because otherwise it interferes with the user deletion
+                remove_user_from_represented_and_ccing_users(user, deletable_users + users_to_mark_inactive)
```
Comment (janno42) [inline evap/staff/tools.py:193]: you don't show any messages here, are you?

**Review by janno42 — CHANGES_REQUESTED** [review]
thanks! i still have to do some testing, but here is already some feedback. can you also please add automated tests for this new feature?

File: evap/staff/tools.py (hunk 4 above)
Comment (He3lixxx) [reply evap/staff/tools.py:308]: or just use `(user, ignored_users=[], test_run=False)` (keep in mind that the actual list instance is then shared across calls, so you can't do that if you modify the list in the method. Shouldn't be a problem here, though)

Comment (JannisBerndt) [reply evap/staff/tools.py:308]: Using '[]' as default parameter throws an error in the codecov test because of the problem @He3lixxx mentioned (which wozld not be an actual problem im this case). I changed it because of that... The Suggestion of @janno42 seems the best solution then...

File: evap/staff/tools.py (hunk 5 above)
Comment (JannisBerndt) [reply evap/staff/tools.py:193]: No, because they would be doubled with the information messages in the previous for-loops.
Should they be shown twice?
The only real problem is the future tense in the messages, but it is also future in all the other notifications. How should it be done?

**Comment by JannisBerndt** [comment]
> thanks! i still have to do some testing, but here is already some feedback. can you also please add automated tests for this new feature?

Thanks for the feedback, yes I will add some tests.

File: evap/staff/tools.py [hunk 6]
```diff
@@ -292,3 +301,21 @@ def find_next_unreviewed_evaluation(semester, excluded):
...
         .annotate(num_unreviewed_textanswers=Count("contributions__textanswer_set")) \
         .order_by('vote_end_date', '-num_unreviewed_textanswers').first()
+
+
+def remove_user_from_represented_and_ccing_users(user, ignored_users=None, test_run=False):
+    remove_messages = []
+    ignored_users = list() if ignored_users is None else ignored_users
+    for represented_user in user.represented_users.exclude(id__in=[user.id for user in ignored_users]):
+        if test_run:
+            remove_messages.append(_("%s will be removed from the delegates of %s.") % (user.full_name, represented_user.full_name))
```
Comment (He3lixxx) [inline evap/staff/tools.py:311]: We usually the the more recent .format() syntax of python that gets rid of some of the weirdness of the older %-formatting. I'd appreciate if you changed this here (and in the other few cases)

File: evap/staff/tools.py (hunk 5 above)
Comment (janno42) [reply evap/staff/tools.py:193]: it's fine how it is. i didn't see that the "test" messages are also shown on the real update run :)

File: evap/staff/tools.py [hunk 7]
```diff
@@ -292,3 +301,21 @@ def find_next_unreviewed_evaluation(semester, excluded):
...
+
+def remove_user_from_represented_and_ccing_users(user, ignored_users=None, test_run=False):
+    remove_messages = []
+    ignored_users = ignored_users or []
+    for represented_user in user.represented_users.exclude(id__in=[user.id for user in ignored_users]):
+        if test_run:
+            remove_messages.append(_("{} will be removed from the delegates of {}.").format(user.full_name, represented_user.full_name))
+        else:
+            represented_user.delegates.remove(user)
+            remove_messages.append(_("Removed {} from the delegates of {}.").format(user.full_name, represented_user.full_name))
```
Comment (janno42) [inline evap/staff/tools.py:314]: as discussed in a comment above, this message and the one in line 320 are never shown because you're only using them in the bulk update when it's not a test run, correct? then i'd say you can remove the messages altogether.

**Review by janno42 — CHANGES_REQUESTED** [review]
just one more change request from my side :)

File: evap/staff/tools.py (hunk 7 above)
Comment (JannisBerndt) [reply evap/staff/tools.py:314]: The only place where those messages are used is in line 782 in forms.py.
I am not sure whether we should care about different tenses or use one message tense everywhere.
Furthermore I placed this call into the user change form because the delegates and cc users should get removed on every user change. However, displaying the messages is handled individually outside the form (after changing a user)...

Comment (He3lixxx) [reply evap/staff/tools.py:314]: I feel like this method is fine. Having a distinction between "will be changed" for a test run and "was changed" for a real run is a good thing imo. We should change the three calls in bulk_update_users below.

As far as I can see, in `bulk_update_users`, the simple idiomatic snippet:
```python
for user in deletable_users + users_to_mark_inactive:
    messages = remove_user_from_represented_and_ccing_users(user, deletable_users + users_to_mark_inactive, test_run)
    messages.warning(...)
```

should work -- or am I missing something? We may have to make the database transaction a bit bigger, but that shouldn't be much of a problem.

**Review by janno42 — APPROVED** [review]

**Review by He3lixxx — APPROVED** [review]

File: evap/staff/tools.py [hunk 8]
```diff
@@ -170,18 +170,20 @@ def bulk_update_users(request, user_file_content, test_run):
                 create_user_list_html_string_for_message(users_to_mark_inactive)
             )
         )
-    if emails_of_users_to_be_created:
```
Comment (janno42) [inline evap/staff/tools.py:173]: why did you remove this line?

File: evap/staff/views.py [hunk 9]
```diff
@@ -1598,6 +1598,8 @@ def notify_reward_points(grantings, **_kwargs):  # pylint: disable=unused-variab
         form.save()
         delete_navbar_cache_for_users([user])
         messages.success(request, _("Successfully updated user."))
+        for message in form.remove_messages:
+            messages.warning(request, message)
```
Comment (janno42) [inline evap/staff/views.py:1602]: codecov reports line 1602 not being covered by tests. did you miss it or was including it in a test too complex?

**Review by janno42 — CHANGES_REQUESTED** [review]

File: evap/staff/tools.py (hunk 8 above)
Comment (JannisBerndt) [reply evap/staff/tools.py]: Was a mistake, oops...

File: evap/staff/views.py (hunk 9 above)
Comment (JannisBerndt) [reply evap/staff/views.py:1602]: didnt notice because it broke when adding "mocking" to the test. Got fixed in the last commit

**Review by janno42 — APPROVED** [review]

**Comment by janno42** [comment]
Thank you so much for your continuous support! 🚀
